source/contribute.rst

```diff
@@ -41,128 +41,146 @@ contributing back to a Doctrine project.
+  must be available in the minor release that adds the deprecation for
+  the old feature.
+
+With that in mind:
+
+If you are touching something outside the source directory (usually
+``src``, sometimes ``lib``), and that thing is not ``composer.json``,
+you should probably target the lowest branch.
+There is no way that your contribution is going to create a bug for
+the end user, so let us get it on all 3 branches with merges up.
+Examples are: documentation, tests, build scripts, tooling configuration.
+
+If you are touching something inside the source directory, you may
+still contribute to the lowest branch if it is a bug fix, and if you
+are touching comments that are not phpdoc comments.
+
+If the comments are phpdoc comments… it depends. If the phpdoc was
+inaccurate, you should target the lowest branch. If the phpdoc was
+imprecise but technically correct, you should target the next minor
+branch.
+
+If you are doing a refactoring and you think you didn't change the
+behavior of the code, you should still target the next minor branch
+unless the refactoring is necessary for a bugfix. Let us not take any
+unnecessary risks.
+
+If you want to deprecate something, you should target the next minor
+branch: upgrading to a new patch version should not result in any
+additional deprecations.
+
+If you want to add a new feature, you should target the next minor
+branch, because it might have unintentional side effects that break existing code.
+
+If you want to remove or change a public API, you should target the next
+major branch. If you change a public API, you should provide an upgrade
+path for the end user. Ideally, you should make the new API available on
+the next minor branch, deprecate the old API, and remove it in the next
+major.
+
+Have you made your choice? Good. You now need to create a topic branch.
 
-.. code-block:: console
-
-    $ git clone git@github.com:username/dbal.git
-
--  Enter the dbal directory and add the **doctrine** remote
-
-.. code-block:: console
-
-    $ cd dbal
-    $ git remote add doctrine git://github.com/doctrine/dbal.git
-
-Branching from the default branch
----------------------------------
-
-New pull requests are created with the repository's default branch as base branch.
-The default branch is the branch you see when you enter the repository page on GitHub.
-
-.. image:: ../images/default-branch.png
-   :alt: The default branch
-   :style: margin-bottom: 20px
-
-In this DBAL example, it's the branch with the name **2.11.x**. The branch name reflects the
-current lowest supported version of a repository.
-
-Newly introduced changes to 2.11.x will be up-merged at a later point in time
-to newer version branches (e.g. 2.12.x, 3.0.x). This way you don't have to
-re-introduce a new fix or feature of 2.11.x with another pull request to
-the other version branches.
-
-Keeping the default branch up-to-date!
---------------------------------------
+Creating a topic branch
+-----------------------
```

Comment:
Same as before – do we really need this entire section?

Reply:
For that one I know the answer, and that's yes. There are way too many contributors contributing directly to e.g. 2.14.x directly.

Reply:
Then, let's explain it...

You should avoid making changes directly on the patch, minor or major release branches, also in your fork of the repository. That makes it difficult for you to have more than one pull request at a time, and also complicates matters should you need to rebase your work to another branch some time later on. Instead, for every change that you would like to propose, create a dedicated topic branch. 

Note: If that does not sound all too familiar, you might want to read about GitHub Flow for additional background information https://docs.github.com/en/get-started/quickstart/github-flow.

Reply:
✅